### PR TITLE
Go Language Support – Milestone v0.2.0

### LINKED ISSUE #41: Go Support v0.2.0
An issue collecting features and changes for the "Go Support 0.2.0" release:

- CRDT Support
- Eventsoured re-implementation
- Integration Tests for the CRDT implementation, to be used for the TCK
- Integration Tests for the Eventsourced implementation
- Adoption of new Cloudstate Model based TCK tests
- Protobuf cleanup for go packages
- Eventsourced TCK Support
- Eventsourced effects and forward support
- Significant test coverage increase
- Dapr contrib model support with new CRDT support
- Chat example
- CRDT Shopping Cart example
- `devcontainer.json` Support
- Updated documentation
- CRDT and Effects & Forwards documentation

**CRDT** support has been added as the second state model supported. Together with the Eventsoured implementation, stream runners where introduced which enables multiple entities of the same id to be run at the same time. The first (kind of naive) implementation v0.1.x had a [bug](https://github.com/cloudstateio/go-support/issues/27) where solely one entity of the same id where able to run at the same time.

The **Eventsourced** support was re-implemented using the CRDT stream runner and its way to run a messages stream. Similar, error handing was adopted and also the new atomic command handling error behaviour with entity restart support has been added.

The **Integration Tests for the CRDT** implementation follow the new model-based approach of the Cloudstate TCK. They run with an in-memory gRCP Client and Server and are written in Go. These tests are yet not available in the Cloudstate TCK and complement them having local functional tests and also tests not done by the TCK itself.

The **Protobuf cleanup for go packages** where necessary because the `go_package` used so far, where incompatible and incorrect. This is fixed in this milestone [and also fixed](https://github.com/cloudstateio/cloudstate/issues/455) in the main Cloudstate repository.

With **Eventsoured TCK Support** also **Eventsourced effects and forward support** was correctly implemented and [validated with the TCK](https://observablehq.com/@marcellanz/cloudstate-tck-verification).

A **Significant test coverage increase** [was achieved](https://codecov.io/gh/cloudstateio/go-support/branch/feature%2Fcrdt_support) with about 80 integration- and 150 unit-tests in this milestone.

**Updated documentation** and **CRDT documentation** is done with the new Antora based template.

The new API has been used in different contexts of available examples and public use of cloudstate. One is the **Dapr contrib model support** which uses a CRDT state model. This branch https://github.com/mrcllnz/components-contrib/tree/feature/cloudstate_crdt_support implements the Dapr Cloudstate component with the new Go API.

### PR DESCRIPTION
**Pull request by marcellanz**
Fixes #41

**Features and changes** for the "Go Support 0.2.0" release:
- CRDT Support
- Eventsoured re-implementation
- New API
- Integration Tests for the CRDT implementation, to be used for the TCK
- Integration Tests for the Eventsourced implementation
- Adoption of new Cloudstate Model based TCK tests
- Protobuf cleanup for go packages
- Eventsourced TCK Support
- Eventsourced effects and forward support
- Significant test coverage increase
- Dapr contrib model support with new CRDT support
- Chat example
- CRDT Shopping Cart example
- `devcontainer.json` Support
- Updated documentation
- CRDT and Effects & Forwards documentation

**CRDT** support has been added as the second state model supported. Together with the Eventsoured implementation, stream runners where introduced which enables multiple entities of the same id to be run at the same time. The first (kind of naive) implementation v0.1.x had a [bug](https://github.com/cloudstateio/go-support/issues/27) where solely one entity of the same id where able to run at the same time.

The **Eventsourced** support was re-implemented using the CRDT stream runner and its way to run a messages stream. Similar, error handing was adopted and also the new atomic command handling error behaviour with entity restart support has been added.

The **Integration Tests for the CRDT** implementation follow the new model-based approach of the Cloudstate TCK. They run with an in-memory gRPC Client and Server and are written in Go. These tests are yet not available in the Cloudstate TCK and complement them having local functional tests and also tests not done by the TCK itself.

The **Protobuf cleanup for go packages** where necessary because the `go_package` used so far, where incompatible and incorrect. This is fixed in this milestone [and also fixed](https://github.com/cloudstateio/cloudstate/issues/455) in the main Cloudstate repository.

With **Eventsoured TCK Support** also **Eventsourced effects and forward support** was correctly implemented and [validated with the TCK](https://observablehq.com/@marcellanz/cloudstate-tck-verification).

A **Significant test coverage increase** [was achieved](https://codecov.io/gh/cloudstateio/go-support/branch/feature%2Fcrdt_support) with about 80 integration- and 150 unit-tests in this milestone.

**Updated documentation** and **CRDT documentation** is done with the new Antora based template.

The new API has been used in different contexts of available examples and public use of cloudstate. One is the **Dapr contrib model support** which uses a CRDT state model. This branch https://github.com/mrcllnz/components-contrib/tree/feature/cloudstate_crdt_support implements the Dapr Cloudstate component with the new Go API.

With the **New API** of this milestone, the API interfacing cloudstate instances as well as handling state model have changed slightly. For event sourcing, one is the removal of the `cloudstate.EventEmitter` type to be embedded with every entity, instead the `eventsourced.EntityHandler` has to be implemented providing `HandleCommand` and `HandleEvent` methods. Embedding a type as part of an API seemed a bit too awkward. With this change, the Go `context.Context` typed parameter of command and event handlers have been changed to `eventsource.Context` that embeds a `context.Context`. This simply prevents an unfortunate API like

```go
func (sc *ShoppingCart) AddItem(ctx context.Context, c *eventsourced.Context, li *AddLineItem) (*empty.Empty, error)
```
where two context parameters are provided, one most of the time never used. Instead, the `context.Context` context is embedded and available through the API. Embedding `context.Context` is discouraged but not completely disallowed. The topic is well known in the Go community and there is work being done to [relax recommendation against putting Contexts in structs](https://github.com/golang/go/issues/22602). The narrative for `context.Context` is basically to discourage sharing state by embedding a context in structs and use it instead, to get the contexts passed along chains of function calls and use them with their purpose. We could have embedded the cloudstate `eventsourced.Context` within a `context.Context` with [context.WithValue](https://golang.org/pkg/context/#WithValue) as other similar projects do it, but I found it too much of a stretch to get them again out again of a `context.Context` context.

Second, the `Snapshotter` and `SnapshotHandler` interfaces were combined into the optional `eventsourced.Snapshooter` interface.

Regarding error handling, the in Cloudstate otherwise well known `context.Fail` method to be used to signal the failure of a running context, was replaced with the in Go idiomatic return of an error as a value to be returned explicitly. This transformed the following kind of awkward code snipped:
```go
func (sc *ShoppingCart) AddItem(ctx *eventsourced.Context, li *AddLineItem) (*empty.Empty, error) {
	if li.GetQuantity() <= 0 {
                ctx.Fail(fmt.Errorf("cannot add negative quantity of to item %q", li.GetProductId()))
		return nil, nil
	}
        // ...
}
```
into:
```go
func (sc *ShoppingCart) AddItem(ctx *eventsourced.Context, li *AddLineItem) (*empty.Empty, error) {
	if li.GetQuantity() <= 0 {
		return nil, fmt.Errorf("cannot add negative quantity of to item %q", li.GetProductId())
	}
        // ...
}
```
which is much more explicit as it embraces the principle of [Errors are values](https://blog.golang.org/errors-are-values) in Go.

Regarding the API of CRDT types, a decision was made how to represent, or better said encode, values in structures like maps, sets and registers. We follow the [interface{} says nothing](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=7m36s) proverb, where we simply don't put empty interface typed values into a map, set or register. Until probably [2021](https://blog.golang.org/generics-next-step), when Go gets generics-support, we put `*any.Any` values into maps, sets and registers. There is no magic of how to find out what it is in a map, the type used for `ORMaps` and `ORSets` as well as `LWWRegisters`. Go maps actually are generic types, where the compiler generates code for declared types like `map[string]*shopping.Cart` but this is not available in a dynamic way like we need it for our maps, sets and registers. While `*any.Any` types are kind of like empty interfaces they are more explicit that anything else in the context of Cloudstate, Cloudstate Go users know that they implement gRPC services and that their representation have a relationship to be`*any.Any` types.

Second, as we put `*any.Any` values into maps, sets and registers, we have to decide when to marshal (encode) values into  `*any.Any` values. We could have chosen to encode values at times when cloudstate protocol replies where prepared to be sent (ActionReplies, CRDT State and Deltas) or right when they're inserted into CRDT types. I've chosen the approach to [fail fast](https://wiki.c2.com/?FailFast) (thanks Joe) and put the burden to the API user. It's not that much of a hassle I think, it also shows how explicit we use and implement state-model as we don't hide them (actually we can't) inside an implementation of a `Set` or `Map` with then some implicit or explicit restrictions other Language Support libraries (rightfully in their field) have done.

There are occasions where this decision brings an additional decode and probably encode step like the `getCart` method in the CRTD shopping cart example:

```go
if err := encoding.DecodeStruct(state.GetLwwregister().GetValue(), &item); err != nil {
     return nil, err
}
```
as we have to get values out of a CRDT type again and, to return them in the form of a gRCP service return type. If performance-critical, this can be explicitly optimized. The next iteration of the support library will show if that was the right decision or if it has to be adjusted. I did not want to optimize without empirical data. Also, `any.Any` types for many Cloudstate state model types are the encoded format that gets to be known for the proxy. This said, the second Go proverb chosen in this issue that says: [Clear is better than clever](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=14m35s) is reflected and implemented in this PRs work.

### CODE REVIEW

**Comment by sleipnir** [comment]
Really great works @marcellanz

**Comment by pvlugter** [comment]
:tada: 🚀 Very cool! Awesome work.

**Comment by ralphlaude** [comment]
Awesome @marcellanz!!

**Comment by jboner** [comment]
Wow. Well done, @marcellanz, well done.

File: cloudstate/discovery/server.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+		if err != nil {
+			return err
+		}
+	}
+	return s.updateSpec()
+}
+
+func (s *EntityDiscoveryServer) registerFileDescriptor(msg descriptor.Message) error {
+	fd, _ := descriptor.ForMessage(msg) // this can panic.
+	if r := recover(); r != nil {
```
Comment (chongkai) [inline cloudstate/discovery/server.go]: recover has to be in a defer block

**Review by chongkai — CHANGES_REQUESTED** [review]

File: cloudstate/discovery/server.go (hunk 1 above)
Comment (marcellanz) [reply cloudstate/discovery/server.go]: @chongkai great catch!
We even can let it panic I think, as `descriptor.ForMessage` does not document that it crashes at all.

The deprecated use of `descriptor.ForMessage` will be mitigated with the switch over to protobuf API v2 with https://github.com/cloudstateio/go-support/issues/20

File: cloudstate/cloudstate.go [hunk 2]
```diff
@@ -16,230 +16,92 @@
...
 	"google.golang.org/grpc"
 )
 
-const (
-	SupportLibraryVersion = "0.1.1"
-	SupportLibraryName    = "cloudstate-go-support"
-)
-
-// CloudState is an instance of a CloudState User Function
+// CloudState is an instance of a Cloudstate User Function.
```
Comment (sleipnir) [inline cloudstate/cloudstate.go:33]: CloudState is an instance of a CloudState User Function.?

Comment (marcellanz) [reply cloudstate/cloudstate.go:33]: nice catch; do you refer to the word casing or the statement itself?

Comment (marcellanz) [reply cloudstate/cloudstate.go:33]: Ok, this time `cloudstate.CloudState` is the Go type representing a Cloudstate user function instance. As Cloudstate, with lower case state, is the marketing name, it felt strange to name Go types using a marketing name and not CamelCase naming the type.

Actually we could name `cloudstate.CloudState` `cloudstate.Server` which would escape us from the unholy duplication of naming a type respecting the package to be part of the type named. I thought about that before, but did not yet was convinced to do it.

Comment (sleipnir) [reply cloudstate/cloudstate.go:33]: Perfect!

**Comment by pvlugter** [comment]
:tada:

**Pull request merged**